### ISSUE TITLE
Resampling block

### ISSUE THREAD

**Issue by sdatkinson** [sdatkinson/NeuralAmpModelerCore#59]
Implement `Resample`, which subclasses `DSP`. It encapsulates another `DSP` object, and takes care of the incoming and outgoing resampling required to expose a buffer to the encapsulated object at the sample rate that it expects.

The use case for this is https://github.com/sdatkinson/NeuralAmpModelerPlugin/issues/59. For example, a NAM model that expects a sample rate of 48k may be encapsulated, and the DAW may be running the plugin with a buffer provided at a sample rate of 44.1kHz.  `Resample` will:
* Resample the incoming buffer to 48k
* Provide the resampled buffer to the encapsulated NAM model
* Resample the NAM model's output back to 44.1k.

This will incur a delay due to the bounding samples that are required to interpolate a given sample as well as the integer sample numbers, meaning that a varying number of samples will be outputted by both of the resampling functions.

If the inner and outer sampling rates are equal, then it would be great to bypass any interpolation math and do a straight pass-through.

The purpose of the resampling block is to expose a clean interface around the encapsulated DSP object, and should be orthogonal from the resampling algorithm itself that's used. I'll provide a simple cubic interpolation algorithm, and other developers may swap in more accurate or better-optimized routines as they'd like.

**Comment by daleonov** [comment]
Yes! You can look at the resampler from JUCE for inspiration. I've been using it with NAM since last summer and it works perfectly for this exact purpose. So might as well port it to NAM. I use a ```LagrangeInterpolator``` from here -> https://docs.juce.com/master/classGenericInterpolator.html

**Comment by daleonov** [comment]
Oh, by the way, two more requests for that:
1. ```getLatency()``` method that tells us how many frames is the resampler's latency, so we can report it to the DAW to adjust PDC.
2. ```prepare()``` method to call outside audio thread that resizes those internal buffers every time sample rate or block size changes.

(obviously method names are arbitrary)

**Comment by David-Else** [comment]
A few years ago this was considered the highest quality open source resampling library from the master behind Voxengo plugins: https://github.com/avaneev/r8brain-free-src. I assume it still is, check it out!

**Comment by olilarkin** [comment]
I am not sure that the [cubic resampler WIP](https://github.com/sdatkinson/NeuralAmpModelerCore/blob/885a535d29022138f3efcbfd73aed4743f750b68/dsp/Resample.h#L38) in the NeuralAmpModelerCore repo will be of high enough quality. Cubic resampling will introduce a lot of aliasing. But r8brain-free-src is not designed for real time usage afaik, and will be too heavy. I propose to use either [WDL_Resampler](https://github.com/justinfrankel/WDL/blob/main/WDL/resample.cpp) or perhaps [this](https://github.com/surge-synthesizer/sst-basic-blocks/blob/main/include/sst/basic-blocks/dsp/LanczosResampler.h) which @baconpaul kindly MIT'd for me.

**Comment by olilarkin** [comment]
Implementing [WDL_Resampler](https://github.com/justinfrankel/WDL/blob/main/WDL/resample.cpp) in the plug-in code would be simple

**Comment by olilarkin** [comment]
I've got some really good results with the Lancsoz resampler, which has much less aliasing that the interpolation modes and is fast and only adds a tiny bit of latency: https://github.com/sdatkinson/NeuralAmpModelerPlugin/pull/341

**Comment by baconpaul** [comment]
Once I fixed that bug. Lol. Glad to see it helps!

**Comment by sdatkinson** [comment]
Coming back to this, I don't think it makes sense to implement the resampler here: same as [AudioDSPTools](https://github.com/sdatkinson/AudioDSPTools), it's not "NAM-specific" and others may have their own favorites already.

Still, I definitely appreciate the input on this Issue and will be referencing it as I work on getting resampling in my plugin; thanks!
